README.md
# NAME

Call::Haskell - Call Haskell functions from Perl

# SYNOPSIS

    use Call::Haskell 'My::Haskell::Module( f1, f2, f3, f4 )'

or more explicitly:

    use Call::Haskell import => 'My::Haskell::Module( f1, f2, f3, f4 )' , path => '..', clean => 0, verbose => 0 ;

    my $res = f1(@args);

# DESCRIPTION

`Call::Haskell` provides a simple mechanism to call Haskell functions from Perl. The syntax for the import statement is similar to what you would write in Haskell. If the path is not specified, the local directory is assumed.

To use this module you need

    - perl 5.16 or later,
    - ghc 7.8 or more recent,
    - gcc 4.8 or more recent

You also need recent versions of

    - the Inline, Inline::C and Digest::MD5 Perl packages
    - the Parsec Haskell package

The Haskell function arguments and return values must have types that are lists, tuples or maps of primitive types (`Int`, `Bool`, `String`). Haskell's `Data.Map` becomes a Perl hash and vice versa. `Maybe` is also supported, `Nothing` is mapped to `undef` and vice versa.

If this is too restrictive, you can use the `Functional::Types` module which provides a Haskell-like type system for Perl. In this case, your corresponding Haskell types must be instances of the `Typeable` typeclass.

The module packs the arguments into a string and unpacks the return value from a string. This will be very slow for large data structures.

Currently, you can only use functions from a single Haskell module.

The module creates two subdirectories in your working directory: `_Call_Haskell` and `_Inline`. You can find all generated code in there.

# TESTING

The Haskell build process does currently not work with automated testing so the test in `t/basic.t` is a stub. You can test manually by commenting out the current plan, and uncommenting the lines below it:

    #plan skip_all => "Skipping all test, please uncomment the 'use Call:Haskell' and the test plan and run manually";
    plan tests => 4;
    use Call::Haskell import => 'ProcessStr( f1, f2, f3, f4 )', path => '.';

Then you can test it as follows:

    $ cd t
    $ perl -I../lib basic.t

# AUTHOR

Wim Vanderbauwhede <Wim.Vanderbauwhede@mail.be>

# COPYRIGHT

Copyright 2015- Wim Vanderbauwhede

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO

- Inline, Inline::C and Digest::MD5 Perl packages
- The Parsec Haskell package <http://hackage.haskell.org/package/parsec>
- A presentation about `Call::Haskell` and `Functional::Types`: <http://www.slideshare.net/WimVanderbauwhede/perl-andhaskell>


Changes
Revision history for Call-Haskell

{{$NEXT}}

v0.2.1  2015-09-18 13:25:18 BST
        - Localised Data::Dumper settings
        - Added instructions for manual testing to README.md
v0.2.0  2015-03-10 15:52:21 GMT
        - Builder now correctly detects objects to be put in the FFI library
        - Builder now automatically detects Haskell packages for linking

v0.1.1  2015-02-19 10:06:24 GMT
        - Changed name of types module from Types to Functional::Types

v0.1.0  2015-02-18 09:17:57 GMT
        - This release works for most untyped cases as well as with the type system.

v0.0.1  2015-02-09 14:34:36 GMT
        - Initial release


lib/Call/Haskell/ReadShow.pm
package Call::Haskell::ReadShow;
use version; our $VERSION = version->declare('v0.1.0');
use warnings;
use strict;
use v5.16;
# Uses ShowToPerl to serialise the output of the Haskell function

use Data::Dumper;

# This routine needs the type signature from the Haskell function.
sub showH { (my $data, my $hs_type)=@_;
    my $t=$hs_type;
    if(ref($t) eq 'ARRAY' && scalar( @{$t}) ==1) { # List 
        if ($hs_type->[0] eq 'Char') { # String
            return show_string($data); 
        } else {
            my $list_t= $t->[0];
            return show_list($data, $list_t);
        }
    } elsif (ref($t) eq 'ARRAY' && scalar( @{$t})>1) { # Tuple            
            return show_tuple($data, $t);
  } elsif (ref($t) eq 'HASH' ) { # Alg. Datatype            
            return show_algtype($data, $t);            
    } elsif ($t =~/[A-Z]/) { # Either prim, Map or Maybe
        if($hs_type =~/^([A-Z]\w*)$/) { # Prim, which means Int, Integer, Float, Double, Integral, Fractional etc or Bool. 
            my $prim_t=$1;
            return show_prim($data,$prim_t);
        } else {
            die "Something wrong (2) with the type ".Dumper($hs_type)."!\n"; 
        }
    } else {die "Something wrong (1) with the type ".Dumper($hs_type)."!\n" }
}

sub show_string { '\"'.$_[0].'\"' };

sub show_prim {
        (my $data, my $hs_type)=@_;
        if ($hs_type eq 'Bool') {
            return ($data ? 'True' : 'False')
        } else {
            return $data;
        }
}

sub show_maybe {
    (my $data, my $hs_type)=@_;
    if (defined $data) {
        return 'Just '.showH($data,$hs_type);
    } else {
        return 'Nothing';
    }
}

sub show_list {
    (my $data, my $hs_type)=@_;
    return '['.join(',', map { showH($_,$hs_type) } @{$data}).']';
}

sub show_tuple {
    (my $data, my $hs_type)=@_;
    my @hs_types=@{$hs_type}; 
    if (scalar @{$data} != scalar @hs_types) {die "Tuple type and values do not match!\n" };    
    return '('.join(',', map { showH($_,shift(@hs_types)) } @{$data}).')';
}

sub show_hash {
    (my $data,my $key_t,my $val_t)=@_;
    my @kv_list=();
    while ((my $k, my $v) = each(%{$data})) {
        push @kv_list,'('.showH($k,$key_t).','.showH($v,$val_t).')';
    }
    return 'fromList ['.join(',',@kv_list).']';
}

sub show_algtype {
    (my $data, my $hs_type)=@_;
    if ($hs_type->{'TypeName'} eq 'Map') {
        (my $key_t,my $val_t) = @{$hs_type->{'TypeArgs'}};
        show_hash($data,$key_t,$val_t);
    } elsif ($hs_type->{'TypeName'} eq 'Maybe') {
        show_maybe($data, $hs_type->{'TypeArgs'}->[0] );
    } else {
        # This is an unknown datatype. Give up
         die "Type ".  $hs_type->{'TypeName'}. " not supported, please create a Perl counterpart using the Functional::Types module\n";
    }
}

# Simply eval the string from ShowToPerl, using AUTOLOAD to clean up.
sub readH {
    (my $vts)=@_;
    my $ref=eval( $vts );
    return $ref;
}


sub AUTOLOAD {
    our $AUTOLOAD;
        my $t=$AUTOLOAD;
        $t=~s/^.+:://;
        $t eq 'True' && do{$t=1};
        $t eq 'False' && do{$t=1};
        $t eq 'Nothing' && do{$t=undef};
    if (not @_) {
        return $t;
    } else {
            return {TypeName=>$t,TypeArgs=>[@_] };
    }
}

1;


lib/Call/Haskell/FFIGenerator.pm
package Call::Haskell::FFIGenerator;
use warnings;
use strict;
use v5.16;
use Data::Dumper;
use Cwd;
use Digest::MD5;

use version; our $VERSION = version->declare('v0.2.0');

use Exporter 'import';

@Call::Haskell::FFIGenerator::EXPORT = qw(
  create_hs_ffi_generator
);

# This module generates all the glue code needed for calling Haskell from Perl.
# What we should do is test if the original module has changed via a checksum
our $VV=0;
sub create_hs_ffi_generator {
 ( my $module_name, my $function_names, my $inc, my $CLEAN, my $VV, my $perl_types ) = @_;
 $Call::Haskell::FFIGenerator::VV=$VV;
 
 my $wd = cwd();
 my $Call_Haskell_path=$INC{"Call/Haskell/FFIGenerator.pm"};
 my $hs_FFIGenerator_dir=$Call_Haskell_path;

 $hs_FFIGenerator_dir=~s/Call.Haskell.FFIGenerator.pm$//;
 if ($hs_FFIGenerator_dir=~/^\.\./) {
     $hs_FFIGenerator_dir='../'.$hs_FFIGenerator_dir;
 }
 if ( not -e '_Call_Haskell' ) {
  mkdir '_Call_Haskell';
 }
 chdir '_Call_Haskell';
 mkdir 'FFIGenerator';
 system("cp $hs_FFIGenerator_dir/FFIGenerator/*.hs FFIGenerator");
 my $generate = 0;
 if ( $CLEAN == 1 ) {
#  say "CLEANING generated sources" if $VV;
#  unlink $module_name . '_ffi_wrapper_gen.hs';
#  unlink $module_name . "FFIWrapper.hs";
#  unlink $module_name . "CWrapper.c";
#  unlink $module_name . "CWrapper.h";    
  $generate = 1;
 } else {
#  say "Computing MD5 for $wd/$inc/$module_name.hs";
  my $filename = "$wd/$inc/$module_name.hs";
  open( my $fh, '<', $filename ) or die "Can't open '$filename': $!";
  binmode($fh);
  my $md5 = Digest::MD5->new;
  while (<$fh>) {
   $md5->add($_);
  }
  close($fh);
  my $checksum = $md5->hexdigest;
  if ( -e "$module_name.md5" ) {
   open( my $md5fh, '<', "$module_name.md5" );
   my $ref_checksum = <$md5fh>;
   close $md5fh;   
   say "MD5:", $checksum eq $ref_checksum if $VV;
   if ( $checksum ne $ref_checksum ) {
    unlink "$module_name.md5";    
    $generate = 1;
   } else {       
    $generate = 0;
   }
  } else {
      say "no MD5 checksum for $module_name.md5" if $VV;
      
   $generate = 1;
  }
  if ( not -e "$module_name.md5" ) {
   $generate = 1;
   
   open( my $md5fh, '>', "$module_name.md5" );
   print $md5fh $checksum;
   close $md5fh;
  }
 }
 if (
  not( -e $module_name . '_ffi_wrapper_gen.hs'
   and -e $module_name . "FFIWrapper.hs"
   and -e $module_name . "CWrapper.c"
   and -e $module_name . "CWrapper.h"
   and -e $module_name . "_Inline_C.c" )
   )
 {

  $generate = 1;
 }
my $c_code='';
 if ( $generate == 1 ) {

  my $hs_module_dir = $inc;

  open my $GEN, '>', "${module_name}_ffi_wrapper_gen.hs";
  say $GEN "-- Code generated by $0 on " . scalar localtime;
  while ( my $line = <DATA> ) {
   if ( $line =~ /_MODULE_NAME_/ ) {
    $line =~ s/_MODULE_NAME_/$module_name/;
   }
   if ( $line =~ /_FUNC_/ ) {
    my $nlines = '';
    for my $function_name ( @{$function_names} ) {
     my $nline = $line;
     $nline =~ s/_FUNC_/$function_name/g;
     $nlines .= $nline;
    }
    $nlines =~ s/\,$//;
    $line = $nlines;
   }
   print $GEN $line;
  }
  close DATA;
  close $GEN;
 my $Call_Haskell_path=$INC{"Call/Haskell/FFIGenerator.pm"};
 my $hs_FFIGenerator_dir=$Call_Haskell_path;

 $hs_FFIGenerator_dir=~s/Call.Haskell.FFIGenerator.pm$//;

 #if ($hs_FFIGenerator_dir=~/^\./) {die "\nThe path to the Call::Haskell module _must_ be absolute, please redefine your PERL5LIB\n\n"; }
 $hs_FFIGenerator_dir='.'; 
 say  "FFIGenerator path:",$hs_FFIGenerator_dir if $VV;   
   my @hs_f_types = 
    `runhaskell -i$wd/$hs_module_dir -i$hs_FFIGenerator_dir ${module_name}_ffi_wrapper_gen.hs`;
   $c_code = create_C_wrapper( \@hs_f_types, $module_name );
 
  open my $INL_C_CODE, '>', $module_name . '_Inline_C.c';
  print $INL_C_CODE $c_code;
  close $INL_C_CODE;
  my @decomp_hs_types = map { decomposeHSType($_) } @hs_f_types;
#  say Dumper(@hs_f_types);
#  die Dumper(@decomp_hs_types);
  mkdir "$wd/_Call_Haskell/CallHaskellWrappers";
  for my $ft (@decomp_hs_types) {
   (my $fn, my $arg_types, my $ret_type)=@{$ft};
   if(isSimpleSig(@{$arg_types}, $ret_type)==0) {   
      if (not -e "$wd/_Call_Haskell/CallHaskellWrappers/$fn.pm")  {
       # create the sub       
       my $code = <<"ENDWC";
package CallHaskellWrappers::$fn;
use Exporter qw( import );
\@CallHaskellWrappers::${fn}::EXPORT = qw( $fn );

ENDWC
    my $hs_argtup_type = (scalar @{$arg_types} > 1) ?  '('.join(',',@{$arg_types}).')' : $arg_types->[0];

    local $Data::Dumper::Indent=0;  
    local $Data::Dumper::Terse=1;
        
    my $hs_type_str=Dumper($hs_argtup_type);
    if ($perl_types eq '') {             
        $code .= <<"ENDWH";
use Call::Haskell::ReadShow qw( showH readH );
require Call::Haskell; 
sub $fn {
    my \$hs_type=$hs_type_str;
    my \$in_str = Call::Haskell::ReadShow::showH(\@_,\$hs_type);
    my \$out_str=Call::Haskell::${fn}_ser(\$in_str);
    my \$res = Call::Haskell::ReadShow::readH(\$out_str);
    return \$res;
}

ENDWH
        
       } else {
my $use_perl_types = "$perl_types" eq '1' ? '' : "use $perl_types;";
    
       $code .= <<"ENDWTP";
use Call::Haskell::ReadShow qw( showH readH );       
use Types;
$use_perl_types
require Call::Haskell; 
sub $fn {
    my \$hs_type=$hs_type_str;
#    my \$in_str = '';
    my \@in_arg_strs=();
    for my \$arg (\@_) {
        if (ref(\$arg) eq 'Types') {    
            push \@in_arg_strs, Types::show(\$arg);
        } else {
           push \@in_arg_strs,  Call::Haskell::ReadShow::showH(\$arg,\$hs_type);
        }
    }
    my \$in_str = (\@_>1) ? '('.join(', ',\@in_arg_strs).')' : \$in_arg_strs[0];     
    my \$out_str=Call::Haskell::${fn}_ser(\$in_str);    
    my \$res = eval(\$out_str);
    return \$res;
}

ENDWTP

$code .= "1;\n";
       }
open my $SUB,'>',"$wd/_Call_Haskell/CallHaskellWrappers/$fn.pm";
print $SUB $code;
close $SUB;
      }
   }
    }   
 }
 else {
  say "NO GENERATION NECESSARY!" if $VV;
  open my $INL_C_CODE, '<', $module_name . '_Inline_C.c';
  while ( my $line = <$INL_C_CODE> ) {
   $c_code .= $line;
  }
  close $INL_C_CODE;
 }
 
 return ( $c_code, $generate );
}

my %SimpleTypes =
  map { ( $_ => 1, "IO $_" => 1 ) }
  qw( Int Integer Float Double String [Char] );

sub HstoP_types {
 ( my $hs_t, my $is_retval ) = @_;
 my $p_t = 'NONE';
 if ( $hs_t eq 'Int' ) {
  $p_t = 'long';
 }
 elsif ( $hs_t eq 'Double' ) {
  $p_t = 'double';

 }
 elsif ( $hs_t eq '[Char]' or $hs_t eq 'String' ) {
  if ($is_retval) {
   $p_t = 'SV*'
     ; # SvPV(sv_name, PL_na)  char* n_c_str = newSVpvf("STR: <%s %s> OK\n", c_str,hs_str);
  }
  else {
   $p_t = 'char*';
  }
 }
 else {

  # This type needs to be serialised, so it's a string
  # This is not correct on a per-arg basis, but we should not come here
  if ($is_retval) {
   $p_t = 'SV*'
     ; # SvPV(sv_name, PL_na)  char* n_c_str = newSVpvf("STR: <%s %s> OK\n", c_str,hs_str);
  }
  else {
   $p_t = 'char*';
  }
 }
 return $p_t;
}
# So with the new type datastructure this will work if the type is a string and it exists. 
# So that's fine in a very Perl-ish way 
sub isSimpleSig {
 ( my @hs_f_types ) = @_;
 my $test = 1;
 map { $test = $test && ( exists $SimpleTypes{$_} ? 1 : 0 ) }
   @hs_f_types  ;#( @{$argtypes}, $ret_type );
# say "isSimpleSig <$test>";
 return $test;
}

sub decomposeHSTypeOLD {
 ( my $hs_f_type ) = @_;
# say "decomposeHSType:",$hs_f_type;
 chomp $hs_f_type;
 ( my $f, my $args ) = split( /\s+::\s+/, $hs_f_type );
 my @argtypes = split( /\s+->\s+/, $args );
 my $ret_type = pop @argtypes;
 my $res= [ $f, \@argtypes, $ret_type ];
# say Dumper($res);
 return $res;
}

sub decomposeHSType {
 ( my $hs_f_type ) = @_;
# say "decomposeHSType:",Dumper($hs_f_type);
 chomp $hs_f_type;
 ( my $f, my $args ) = split( /\s+::\s+/, $hs_f_type );
# say "decomposeHSType:",$args;
 my $argtypes = eval($args);
 my $ret_type = pop @{$argtypes};
 my $res= [ $f, $argtypes, $ret_type ];
# say Dumper($res);die;
 return $res;
}

sub create_C_wrapper {
 ( my $hs_f_types, my $module_name ) = @_;
 my $inc    = '#include "' . $module_name . 'CWrapper.h"';
 my $date   = scalar localtime;
 my $c_code = "// Code generated by $0 on $date
$inc
int hs_begin(int n) {
    hs_${module_name}_init();
    return n;
}

int hs_end(int n) {
    hs_${module_name}_end();
    return n;
}
";
 my $ws = '    ';
 for my $hs_f_type ( @{$hs_f_types} ) {
  ( my $f, my $argtypes, my $res_t ) = @{&decomposeHSType($hs_f_type)};
  
#  say "TYPE: $hs_f_type";
#  say "DECTYPE:", Dumper($argtypes);
#  say $f;
  my @fdef_lines = do {

   if ( isSimpleSig(@{$argtypes}, $res_t) == 1 ) {
    say "SIMPLE!" if $VV;
    my $i = 0;
    (
     '',
     HstoP_types( $res_t, 1 ) . " $f("
       . join( ',', map { HstoP_types( $_, 0 ) . ' x' . ++$i } @{$argtypes} )
       . ') {',

     #            $ws.'printf("BEFORE FFI CALL\n");',
     $ws
       . HstoP_types( $res_t, 0 )
       . " res = ${f}_ffi_c("
       . join( ', ', map { "x$_" } ( 1 .. $i ) ) . ');',

     #            $ws.'printf("AFTER FFI CALL\n");',
     $ws 
       . HstoP_types( $res_t, 0 ) 
       . ' nres = '
       . (
      ( $res_t eq '[Char]' or $res_t eq 'String' ) ? 'newSVpv(res,0);' : 'res;'
       ),

     $ws . 'return nres;',
     '}',
     ''
    );
   }
   else {
    say "SERIALISE!" if $VV;
    (
     '',
     "SV* ${f}_ser(char* sstr) {",
     $ws . "char* res = ${f}_ffi_c(sstr);",
     $ws . 'SV* nres = newSVpv(res,0);',
     $ws . 'return nres;',
     '}', ''
    );
   }
  };

  my $fdef = join( "\n", @fdef_lines );

  #    print $fdef;
  $c_code .= $fdef;
 }
 return $c_code;
} # END of create_C_wrapper()

sub AUTOLOAD {
    our $AUTOLOAD;
        my $t=$AUTOLOAD;
        $t=~s/^.+:://;
        $t eq 'True' && do{$t=1};
        $t eq 'False' && do{$t=1};
        $t eq 'Nothing' && do{$t=undef};
    if (not @_) {
        return $t;
    } else {
            return {TypeName=>$t,TypeArgs=>[@_] };
    }
}

1;

# _FUNC_ gets populated for every function with magic commas

__DATA__
module Main where
import Data.Typeable ( typeOf )
-- import Data.List ( intercalate )
import FFIGenerator.TypeToPerl (  typeToPerl ) 
import FFIGenerator.GetTypes ( getFFITypes, hasSimpleSig, getTypes )
import FFIGenerator.GenerateCode ( createHaskellWrapper, createCWrapper, createCWrapperHeader )

import _MODULE_NAME_ 
-- ( 
--         _FUNC_,
--         )

module_name = "_MODULE_NAME_"
func_list =[
    ("_FUNC_", getFFITypes _FUNC_ , hasSimpleSig _FUNC_ , getTypes _FUNC_ ),
  ]

-- So this generated module provides module_name and func_list 
-- To do so, it requires getFFITypes and hasSimpleSig from FFIGenerator.GetTypes
-- The functions that generate the code can go in FFIGenerator.GenerateCode
 
main = do
    createHaskellWrapper module_name func_list
    createCWrapper module_name func_list
    createCWrapperHeader module_name func_list
    putStrLn $ "_FUNC_ :: "++(typeToPerl (show(typeOf _FUNC_)))
--    mapM (\(f,ffit,fs,ft) -> putStrLn (f++" :: "++ (intercalate " -> " ft))) func_list


lib/Call/Haskell.pm
package Call::Haskell;
use warnings;
use strict;
use v5.16;

use Call::Haskell::FFIGenerator qw( create_hs_ffi_generator );
use Cwd;
use Config;
require Inline;

use version; our $VERSION = version->declare('v0.2.0');
@Call::Haskell::ISA = qw(Exporter);
my $VV = 0;


sub import {
 my ( $hs, @import_list ) = @_;

 my %config = (
  'functions' => '',
  'path'       => '.',
  'clean'     => 0,
  'verbose'   => 0,
  'perl_types' => '', 
  'packages' => []
 );
 #TODO: instead of 'packages', we should scan the source code for import statements and then use
 # ghc-pkg find-module $module to find the package
 if ( scalar @import_list == 1 ) {
  $config{'functions'} = $import_list[0];
 }
 else {
  my %import_config = @import_list;
  map { $config{ lc($_) } = $import_config{$_} } keys(%import_config);
 }
 if (exists $config{'import'} ) {
  $config{'functions'}=$config{'import'};
 }
 my $func_import_str = $config{'functions'};

 $func_import_str =~ s/\s+//g;
 $func_import_str =~ s/\)$//;
 my @func_imports = split( /\),/, $func_import_str );
 if ( scalar @func_imports > 1 ) {
  die
"Sorry, at the moment you can only call functions from a single Haskell module\n";
 }
 my $fq_func = shift @func_imports;
 ( my $module, my $funclst ) = split( /\(/, $fq_func );
 my @mfuncs = split( /,/, $funclst );

 #    my %funcs=();
 #    for my $fq_func (@func_imports) {
 #        say "<$fq_func>";
 #        (my $module, my $funclst)=split(/\(/,$fq_func);
 #        my @mfuncs = split(/,/,$funclst);
 #        for my $func (@mfuncs) {
 #        push @{$funcs{$module}},$func;
 #        }
 #    }
 build( $module, \@mfuncs, $config{'path'}, $config{'packages'}, $config{'clean'},
  $config{'verbose'} , $config{'perl_types'} );
  my $wd = cwd();
#  say "Before EXPORT code";
 for my $fn (@mfuncs) {
  if (-e "$wd/_Call_Haskell/CallHaskellWrappers/$fn.pm") {
#   push @Call::Haskell::EXPORT, "${fn}_ser";
  #Call::Haskell->export_to_level( 1, 'Call::Haskell', @Call::Haskell::EXPORT );
  require "$wd/_Call_Haskell/CallHaskellWrappers/$fn.pm";
  eval("CallHaskellWrappers::$fn->import()");
  }
  push @Call::Haskell::EXPORT, $fn;
 }
#say "After EXPORT code, now exporting to level 1";
 Call::Haskell->export_to_level( 1, 'Call::Haskell', @Call::Haskell::EXPORT );
# say "LEAVING Call::Haskell::Import";
}

sub build {
 ( my $hs_module, my $function_names, my $hs_module_dir, my $extra_packages, my $CLEAN, $VV, my $perl_types ) =
   @_;                     #'ProcessString';
 my $wd = cwd();
 if ($CLEAN) {
    system('rm -Rf _Call_Haskell _Inline');
        system('rm *.o *.hi');

 }
 ( my $inline_c_code, my $generated ) =
   create_hs_ffi_generator( $hs_module, $function_names, $hs_module_dir, $CLEAN,
  $VV, $perl_types );

 if ( $generated == 1 ) { $CLEAN = 1 }
 print "INLINE C CODE:\n" . $inline_c_code if $VV;

 if ( not -d './tmp' ) {
  mkdir 'tmp';
 }

 my $perl_link_options_str = $Config::Config{lddlflags};
 my $perl_ld_str           = $Config::Config{ld};
 if ( $perl_link_options_str =~ /-fstack-protector/ ) {
  $perl_link_options_str =~ s/-fstack-protector//;
 }
 print "PERL LD = $perl_ld_str\nPERL LDDLFLAGS = $perl_link_options_str\n"
   if $VV;

 # The code below could go into a module
 #say join(':',@INC),';',"\n",Dumper(%INC);#,';',
 my $Call_Haskell_path=$INC{"Call/Haskell.pm"};

 my $hs_FFIGenerator_dir=$Call_Haskell_path;
 $hs_FFIGenerator_dir=~s/Call.Haskell.pm$//;

 #if ($hs_FFIGenerator_dir=~/^\./) {die "\nThe path to the Call::Haskell module _must_ be absolute, please redefine your PERL5LIB\n\n"; }
 $hs_FFIGenerator_dir='.';
 say  "FFIGenerator path:",$hs_FFIGenerator_dir if $VV;
 my $hs_ffi_module = $hs_module . 'FFIWrapper';

# This requires test_hs_c_str.c to exist and be compiled. I should generate that!
 my $test_src           = 'test_src';
 my $test_out           = 'testl';
 my $c_wrapper          = $hs_module . 'CWrapper';
 my $hs_lib             = $hs_module . 'HsC';
 my $link_options_cache = '.lddlflags.cache';
 my $obj_file_names_cache ='.objs.cache';

# # Clean up
 if ($CLEAN) {
  system("rm *.o *.hi lib$hs_lib.a $test_src.c $test_out $link_options_cache $obj_file_names_cache 2>/dev/null ");
  system("rm -Rf ./tmp/* 2>/dev/null");
 }

 #  Now, first compile the Haskell file:
  my $object_files="";
  my @packages=();
if (not -e  "$obj_file_names_cache") {
# if ( not -e "$hs_ffi_module.o" ) {
  say "ghc -c -O --make -i$wd/$hs_module_dir -i$hs_FFIGenerator_dir $hs_ffi_module" if $VV;
  my @ghc_lines=`ghc -c -O --make -i$wd/$hs_module_dir -i$hs_FFIGenerator_dir $hs_ffi_module`;
  @packages=find_packages(@ghc_lines);
  for my $line (@ghc_lines) {
      chomp $line;
      $line=~s/^.*,\s//; 
          $line=~s/\s\)\s*$/ /;
      $object_files.=$line;
  }
#  say "<<$object_files>>";

  # Here we should write these to a file cache
  open my $OBJS, '>', $obj_file_names_cache;
  print $OBJS $object_files;
  close $OBJS;
 }
 else {
  say "USING CACHED OBJECT NAMES" if $VV;
  open my $OBJS, '<', $obj_file_names_cache;
  $object_files = <$OBJS>;
  close $OBJS;      
# cache the list of object files
 }
 if ( not -e "$c_wrapper.o" ) {
  say "ghc --make -i$wd/$hs_module_dir -i$hs_FFIGenerator_dir -optc-O -no-hs-main -c $c_wrapper.c $hs_ffi_module $hs_module FFIGenerator.ShowToPerl" if $VV; 
  system(
"ghc --make -i$wd/$hs_module_dir -i$hs_FFIGenerator_dir -optc-O -no-hs-main -c $c_wrapper.c $hs_ffi_module $hs_module FFIGenerator.ShowToPerl"
  );
 }
 if ( not -e "lib$hs_lib.a" ) {
  say "ar rcs lib$hs_lib.a  $c_wrapper.o $object_files" if $VV;
#  say "OLD: ar rcs lib$hs_lib.a  $c_wrapper.o $hs_ffi_module.o $wd/$hs_module_dir/$hs_module.o $hs_FFIGenerator_dir/FFIGenerator/ShowToPerl.o";
#  system( "ar rcs lib$hs_lib.a  $c_wrapper.o $hs_ffi_module.o $wd/$hs_module_dir/$hs_module.o $hs_FFIGenerator_dir/FFIGenerator/ShowToPerl.o");
  system( "ar rcs lib$hs_lib.a  $c_wrapper.o $object_files");
 }
 if ( $CLEAN or not -e "$test_src.c" ) {
  my $test_src_code = "#include \"${c_wrapper}.h\"
        int main(int argc, char* argv[]) {
            hs_${hs_module}_init();
            hs_${hs_module}_end();
            return 1;
        }
    ";
  open my $TST, '>', "$test_src.c";
  print $TST $test_src_code;
  close $TST;
  system("gcc -c $test_src.c");
 }
 my $link_options = '';
 my $ld           = '';

 # if there is no cached link options file
 my $package_str = join(' ', map {"-package $_"} (@packages,@{$extra_packages}));
 if ( not -e $link_options_cache ) {
  print
"ghc -v -keep-tmp-files -tmpdir=./tmp  -no-hs-main $test_src.o -L. -l$hs_lib -package parsec -package containers $package_str -o $test_out 2>&1\n"
    if $VV;
  my @ghc_link_output =
`ghc -v -keep-tmp-files -tmpdir=./tmp  -no-hs-main $test_src.o -L. -l$hs_lib -package parsec -package containers $package_str -o $test_out 2>&1`;
  my $ghc_link_options_str = $ghc_link_output[-1];
  $ghc_link_options_str =~ s/^\s*\'//;
  $ghc_link_options_str =~ s/\'\s*$//;
#  print "HASKELL LD CMD: $ghc_link_options_str\n" if $VV;
  my @ghc_link_options = ();
  if ( $ghc_link_options_str =~ /\'\s+\'/ ) {

   @ghc_link_options = split( /\'\s+\'/, $ghc_link_options_str );
  }
  else {
   @ghc_link_options = split( /\s+/, $ghc_link_options_str );
  }
  $ld = shift @ghc_link_options;

  for my $opt (@ghc_link_options) {
   $opt eq '-m64'                 && do { $opt = ''; next };
   $opt eq '-fno-stack-protector' && do { $opt = ''; next };
   $opt eq '-o'                   && do { $opt = ''; next };
   $opt eq $test_out              && do { $opt = ''; next };
   $opt eq '-L.' && do { $opt = '-L../../../../_Call_Haskell'; next };
   $opt eq "$test_src.o"
     && do { $opt = '../../../../_Call_Haskell/' . $opt; next };
   $opt =~ /^tmp.ghc/
     && do { $opt = '../../../../_Call_Haskell/' . $opt; next };
  }
  $ghc_link_options_str = join( ' ', @ghc_link_options );
  $link_options = "$perl_link_options_str $ghc_link_options_str";

  # Here we should write these to a file cache
  open my $LO, '>', $link_options_cache;
  print $LO "$ld :: $link_options";
  close $LO;
 }
 else {
  say "USING CACHED LINK OPTIONS" if $VV;
  open my $LO, '<', $link_options_cache;
  my $ld_link_options = <$LO>;
  close $LO;
  ( $ld, $link_options ) = split( /\s::\s/, $ld_link_options );
 }
 print "=" x 80, "\n" if $VV;
 print "LDDLFLAGS = $link_options\n" if $VV;
 print "=" x 80, "\n" if $VV;
 chdir $wd;
 Inline->import(
  C         => Config => LD => $ld,
  LDDLFLAGS => $link_options,
  'INC'     => "-I$wd/_Call_Haskell"
 );
 Inline->import( C => $inline_c_code );
 say "CALL hs_begin" if $VV;
 hs_begin(1);
# say "LEAVING build()" if $VV;
}

sub find_packages { my @ghc_make_output = @_;
     #TODO: instead of 'packages', we should scan the source code for import statements and then use
 # ghc-pkg find-module $module to find the package
    my @sources=();
    for my $line (@ghc_make_output) {
#        say "LINE: $line";
      chomp $line;
      $line=~s/,.+$//; 
      $line=~s/^.+\(\s//;
#      say "LINE-PROC: $line";
      push @sources,$line;
    }
    my %packages=();
    for my $src (@sources) {
#        say "grep -E '^\\s*import\\s+[A-Z]' $src";
        my @module_lines=`grep -E '^\\s*import\\s+[A-Z]' $src`;
        for my $line (@module_lines) {
#            say "MODLINE: $line";
            my $module = $line;
            $module=~s/^\s*import\s//;
            $module=~s/qualified\s//;
            $module=~s/hiding\s.*$//;
            $module=~s/as\s.*$//;
            $module=~s/\s*\(.*$//;
#            say "MOD: $module";
            my @pkg_lines=`ghc-pkg find-module $module`;
            for my $line (@pkg_lines ) {
                next if $line=~/^\s*$/;
                next if $line=~/:/;
                next if $line=~/\(/;
                $line=~s/^\s+//;
                $line=~s/\-.+\s*$//;
#                say "PKG: <$line>";
                if ($line ne 'base' and $line ne 'mtl') {
                $packages{$line}=1;
                last;
                }
            }
        }    
    }
    delete $packages{'containers'};
    delete $packages{'parsec'};
    return (keys %packages);
}

END {
 say "CALL hs_end" if $VV;
 hs_end(0);
}

1;
__END__

=head1 NAME

Call::Haskell - Call Haskell functions from Perl


=head1 SYNOPSIS

    use Call::Haskell 'My::Haskell::Module( f1, f2, f3, f4 )'

or more explicitly:

    use Call::Haskell functions => 'My::Haskell::Module( f1, f2, f3, f4 )' , path => '..', clean => 0, verbose => 0 ;
    
    my $res = f1(@args);

Note that the path to C<Call/Haskell.pm> must be I<absolute> in C<@INC>. The easiest way is to add the absolute path to the C<PERL5LIB> environment variable.  



=head1 DESCRIPTION

Call::Haskell provides a simple mechanism to call Haskell functions from Perl. The syntax for the import statement is similar to what you would write in Haskell. If the path is not specified, the local directory is assumed.

To use this module you need

    - perl 5.16 or later,
    - ghc 7.8 or more recent,
    - gcc 4.8 or more recent

You also need recent versions of

    - the Inline, Inline::C and Digest::MD5 Perl packages
    - the Parsec Haskell package

Currently the Haskell function arguments and return values must have types that are lists, tuples or maps of primitive types (Int, Bool, String). Haskell's Data.Map becomes a Perl hash and vice versa. Maybe is also supported, Nothing is mapped to undef and vice versa.

The module packs the arguments into a string and unpacks the return value from a string. This will be very slow for large data structures.

Currently, you can only use functions from a single Haskell module.

The module creates two subdirectories in your working directory: C<_Call_Haskell> and C<_Inline>. You can find all generated code in there. 


=head1 AUTHOR

Wim Vanderbauwhede  <Wim.Vanderbauwhede@mail.be>


=head1 COPYRIGHT

Copyright 2015- Wim Vanderbauwhede


=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


=head1 SEE ALSO

=over

=item -

Inline, Inline::C and Digest::MD5 Perl packages


=item -

The Parsec Haskell package L<http://hackage.haskell.org/package/parsec>


=back

